src/rinq/amqp/internal/amqputil/channel_pool.go

```diff
@@ -72,19 +120,90 @@ func (p *channelPool) GetQOS(preFetch uint) (*amqp.Channel, error) {
 }
 
 func (p *channelPool) Put(channel *amqp.Channel) {
+	p.put <- channel
+}
+
+func (p *channelPool) handleGet(request getRequest) error {
+	var response getResponse
+
+	index := len(p.channels) - 1
+	if index >= 0 {
+		// fetch from the pool
+		response.channel = p.channels[index]
+		p.channels = p.channels[:index]
+	} else {
+		// none available, make a new channel
+		response.channel, response.err = p.broker.Channel()
+	}
+	request.reply <- response
+
+	return response.err
+}
+
+func (p *channelPool) handlePut(channel *amqp.Channel) error {
 	if channel == nil {
-		return
+		return nil
 	}
 
 	// set the QoS state back to unlimited, both to "reset" the channel, and to
 	// verify that it is still usable.
 	if err := channel.Qos(0, 0, true); err != nil {
-		return
+		return err
 	}
 
-	select {
-	case p.channels <- channel: // return to the pool
-	default: // pool is full, close channel
+	if len(p.channels) < cap(p.channels) {
+		// return to the pool
+		p.channels = append(p.channels, channel)
+	} else {
+		// pool is full, close channel
 		_ = channel.Close()
 	}
+
+	return nil
+}
+
+func (p *channelPool) handlePeriodicCleanup() {
+	index := len(p.channels) - 1
+	if index >= 0 {
+		// fetch from the pool
+		channel := p.channels[index]
+		p.channels = p.channels[:index]
+		// close channel
+		_ = channel.Close()
+	}
+}
```

Comment:
We need some logic here to determine if the channel has been used.  It may have been removed and put back in the pool 100 times in between ticks, in which case we don't want to close it.

Reply:
Do we want to track every channel's recent time used? I will try with just a single last time since a channel was returned and see how that goes.